### PR TITLE
Provide semantic conventions for RPC metrics

### LINKED ISSUE #1016: Define metric semantic conventions for RPC
**What are you trying to achieve?**
Documented metric semantic conventions for remote procedure calls (RPC).

**What did you expect to see?**
Metric semantic conventions for RPC documented in markdown generated from YAML definitions as per #547.

**Additional context.**
These should be roughly analogous to the trace semantic conventions for RPC [defined here](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/rpc.md). For reference, metric semantic conventions for HTTP operations are already [defined here](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/metrics/semantic_conventions/http-metrics.md).

### PR DESCRIPTION
**Pull request by jsuereth**
Fixes #1016 

## Changes

Adds semantic conventions for RPC metrics.
- Re-use existing RPC attribute spec YAML.  Note: this interacts oddly pulling from trace => metrics.
- Define 3 (x 2) key metrics for RPC:   `rpc.?.duration`, `rpc.?.request_size` and `rpc.?.response_size`, crossed by `server` or `client`
- Add warning around reducing cardinality in attributes for metrics (cross-reference the same information for spans).

Deviations from [OpenCensus](https://github.com/census-instrumentation/opencensus-specs/blob/master/stats/gRPC.md)
- This does not track a "count" of messages that have been issued.
- This does not pull different metrics per-method vs. per-rpc, aggregation across labels is assumed.

### CODE REVIEW

File: specification/metrics/semantic_conventions/rpc.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+*Note: RPC server and client metrics are split to allow correlation across client/server boundaries, e.g. Lining up an RPC method latency to determine if the server is responsible for latency the client is seeing.*
+
+### RPC Server
+
+Below is a table of RPC server metric instruments.
+
+| Name                       | Instrument    | Units        | Description | Status |
+|----------------------------|---------------|--------------|-------------|--------|
+| `rpc.server.duration`      | ValueRecorder | milliseconds | measures duration of inbound RPC | Recommended |
```
Comment (MrAlias) [inline specification/metrics/semantic_conventions/rpc.md:24]: Sorry for the ignorance, but is this a timing of when the request is received to when a response is sent? How would this be used for a streaming RPC?

File: specification/metrics/semantic_conventions/rpc.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+Below is a table of labels that SHOULD be included on metric events and whether
+or not they should be on the server, client or both.
+
+<!-- semconv rpc -->
+| Attribute  | Type | Description  | Example  | Required |
+|---|---|---|---|---|
+| [`rpc.system`](../../trace/semantic_conventions/rpc.md) | string | A string identifying the remoting system. | `grpc`<br>`java_rmi`<br>`wcf` | Yes |
+| [`rpc.service`](../../trace/semantic_conventions/rpc.md) | string | The full name of the service being called, including its package name, if applicable. | `myservice.EchoService` | Conditional<br>No, but recommended |
+| [`rpc.method`](../../trace/semantic_conventions/rpc.md) | string | The name of the method being called, must be equal to the $method part in the span name. | `exampleMethod` | Conditional<br>No, but recommended |
```
Comment (MrAlias) [inline specification/metrics/semantic_conventions/rpc.md:48]: Guessing this was copy/pasted, but we probably do not want to include a reference to a span name here.

```suggestion
| [`rpc.method`](../../trace/semantic_conventions/rpc.md) | string | The name of the method being called. | `exampleMethod` | Conditional<br>No, but recommended |
```

**Review by justinfoote — APPROVED** [review]
This looks good to me. 
My only question is: should the rpc metrics also include HTTP labels if the RPC calls are happening over HTTP?

File: specification/metrics/semantic_conventions/rpc.md (hunk 2 above)
Comment (jsuereth) [reply specification/metrics/semantic_conventions/rpc.md:48]: This is actually how the markdown-generation tool works.   This opens a question on whether we should SHARE rpc attributes between Trace/Metrics.  Right now Trace owns the single-id `rpc` so I just re-used it here because the attributes SHOULD be the same (in my opinion).    However, if we change the documentation for one, we change both (the downside).

So my question to you:

1. Should the attributes be shared?
2. Should I update the tool which generates this markdown from YAML to allow different descriptions?

File: specification/metrics/semantic_conventions/rpc.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+### RPC Server
+
+Below is a table of RPC server metric instruments.
+
+| Name                       | Instrument    | Units        | Description | Status |
+|----------------------------|---------------|--------------|-------------|--------|
+| `rpc.server.duration`      | ValueRecorder | milliseconds | measures duration of inbound RPC | Recommended |
+| `rpc.server.request.size`  | ValueRecorder | bytes        | measures size of RPC request messages | Optional |
+| `rpc.server.response.size` | ValueRecorder | bytes        | measures size of RPC response messages | Optional |
```
[the remark this answers is not in the document]
Comment (jsuereth) [reply specification/metrics/semantic_conventions/rpc.md:26]: Yeah, meant to have these be uncompressed with compressed size as an optional you can also use.

File: specification/metrics/semantic_conventions/rpc.md (hunk 1 above)
Comment (jsuereth) [reply specification/metrics/semantic_conventions/rpc.md:24]: This is the time from when the request was received to when the response is set.

## Streaming RPCs
This likely needs a callout for how to handle streaming RPCs/understand them.

Initially, my thinking is this would be the duration to complete the stream: https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc which doesn't apply in the bi-directional streaming.

I'll update duration to denote its applicability to streaming RPCs and a convention where we can get throughput metrics for streaming RPCs

File: specification/metrics/semantic_conventions/rpc.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
+# General
```
[the remark this answers is not in the document]
Comment (jsuereth) [reply specification/metrics/semantic_conventions/rpc.md:1]: No this mirrors the trace `semantic_conventions` where there are "General RPC" conventions then we provide guidance for specific technologies if necessary (e.g. the `gRPC` callout).

I can rename to "General RPC"

File: specification/metrics/semantic_conventions/rpc.md (hunk 2 above)
Comment (jsuereth) [reply specification/metrics/semantic_conventions/rpc.md:48]: We discussed in spec sig.  I'm going to ask to defer fixing the shared RPC attribute labels to another PR.   There is an existing related proposal (#855), but I'll put forward another proposal that addresses this specific concern and works with existing proposal (or is compatible with it).

**Comment by jsuereth** [comment]
@justinfoote Regarding RPC having HTTP labels if it's over HTTP.  While Personally I could see reasons to include it or not, I think perhaps we should follow from TRACE: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/rpc.md#distinction-from-http-spans

TL;DR:  The metrics is more about the fact that it's an RPC then it is about HTTP GET/POST/PUT/DELETE resources, so we shouldn't cross-report.

File: specification/metrics/semantic_conventions/rpc.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+| Name                       | Instrument    | Units        | Description | Status | Streaming |
+|----------------------------|---------------|--------------|-------------|--------|-----------|
+| `rpc.server.duration`      | ValueRecorder | milliseconds | measures duration of inbound RPC | Recommended | N/A |
+| `rpc.server.request.size`  | ValueRecorder | bytes        | measures size of RPC request messages (uncompressed) | Optional | Recorded per message in a streaming batch |
+| `rpc.server.response.size` | ValueRecorder | bytes        | measures size of RPC response messages (uncompressed) | Optional | Recorded per response in a streaming batch |
+| `rpc.server.requests_per_rpc` | ValueRecorder | count | measures the number of messages received per RPC.  Should be 1 for all non-streaming RPCs | Optional | Required |
+| `rpc.server.responses_per_rpc` | ValueRecorder | count | measures the number of messages sent per RPC.  Should be 1 for all non-streaming RPCs | Optional | Required |
+
+### RPC Client (non-streaming)
```
Comment (yurishkuro) [inline specification/metrics/semantic_conventions/rpc.md:30]: there is a Streaming column in the table

**Review by jmacd — APPROVED** [review]

File: specification/metrics/semantic_conventions/rpc.md (hunk 5 above)
Comment (jmacd) [reply specification/metrics/semantic_conventions/rpc.md:30]: Are you suggestion to remove " (non-streaming)" from the title? (I agree.)

File: specification/metrics/semantic_conventions/rpc.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+*Note: RPC server and client metrics are split to allow correlation across client/server boundaries, e.g. Lining up an RPC method latency to determine if the server is responsible for latency the client is seeing.*
+
+### RPC Server
+
+Below is a table of RPC server metric instruments.
+
+| Name                       | Instrument    | Units        | Description | Status | Streaming |
+|----------------------------|---------------|--------------|-------------|--------|-----------|
+| `rpc.server.duration`      | ValueRecorder | milliseconds | measures duration of inbound RPC | Recommended | N/A |
```
Comment (MrAlias) [inline specification/metrics/semantic_conventions/rpc.md:24]: Does N/A here mean that this instrument is not applicable to a streaming request, as it it shouldn't be used in that situation?

**Review by carlosalberto — APPROVED** [review]

File: specification/metrics/semantic_conventions/rpc.md (hunk 5 above)
Comment (jsuereth) [reply specification/metrics/semantic_conventions/rpc.md:30]: Doh, missed a refactor, thanks!

File: specification/metrics/semantic_conventions/rpc.md (hunk 6 above)
Comment (jsuereth) [reply specification/metrics/semantic_conventions/rpc.md]: I added clarification to why it's N/A.

**Comment by jsuereth** [comment]
@MrAlias Is there anything else I need to do for this PR?

**Review by MrAlias — APPROVED** [review]
LGTM :+1:

**Comment by jsuereth** [comment]
@ carlosalberto What else do I need to do for this PR?

File: specification/metrics/semantic_conventions/rpc.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+| `rpc.client.response.size` | ValueRecorder | bytes        | measures size of RPC response messages (uncompressed) | Optional | Recorded per message in a streaming batch |
+| `rpc.client.requests_per_rpc` | ValueRecorder | count | measures the number of messages received per RPC.  Should be 1 for all non-streaming RPCs | Optional | Required |
+| `rpc.client.responses_per_rpc` | ValueRecorder | count | measures the number of messages sent per RPC.  Should be 1 for all non-streaming RPCs | Optional | Required |
+
+## Labels
+
+Below is a table of labels that SHOULD be included on metric events and whether
+or not they should be on the server, client or both.
+
+<!-- semconv rpc -->
```
Comment (arminru) [inline specification/metrics/semantic_conventions/rpc.md:48]:
```suggestion
```
This is not actually generated by the generator tool (yet), right?

File: specification/metrics/semantic_conventions/rpc.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+| [`net.peer.ip`](../../trace/semantic_conventions/span-general.md) | string | Remote address of the peer (dotted decimal for IPv4 or [RFC5952](https://tools.ietf.org/html/rfc5952) for IPv6) | `127.0.0.1` | See below |
+| [`net.peer.name`](../../trace/semantic_conventions/span-general.md) | string | Remote hostname or similar, see note below. | `example.com` | See below |
+| [`net.peer.port`](../../trace/semantic_conventions/span-general.md) | number | Remote port number. | `80`<br>`8080`<br>`443` | Conditional<br>See below |
+| [`net.transport`](../../trace/semantic_conventions/span-general.md) | string enum | Transport protocol used. See note below. | `IP.TCP` | Conditional<br>See below |
+
+**Additional attribute requirements:** At least one of the following sets of attributes is required:
+
+* [`net.peer.ip`](../../trace/semantic_conventions/span-general.md)
+* [`net.peer.name`](../../trace/semantic_conventions/span-general.md)
+<!-- endsemconv -->
```
Comment (arminru) [inline specification/metrics/semantic_conventions/rpc.md:63]:
```suggestion
```

File: specification/metrics/semantic_conventions/rpc.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
+# General RPC conventions
+
+The conventions described in this section are RPC specific.  When RPC operations
+occur, metric events about those operations will be generated and reported to
+provide insight into those operations.   By adding RPC labels to metric events
+it allows for finely tuned filtering.
+
+**Disclaimer:** These are initial RPC metric instruments and labels but more may
+be added in the future.
+
```
Comment (arminru) [inline specification/metrics/semantic_conventions/rpc.md:10]:
```suggestion
```
This is implicitly assumed for all of our semantic conventions and doesn't need any explicit disclaimer I'd say.

File: specification/metrics/semantic_conventions/rpc.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
+# General RPC conventions
+
+The conventions described in this section are RPC specific.  When RPC operations
+occur, metric events about those operations will be generated and reported to
+provide insight into those operations.   By adding RPC labels to metric events
+it allows for finely tuned filtering.
```
Comment (arminru) [inline specification/metrics/semantic_conventions/rpc.md:6]:
```suggestion
The conventions described in this section are RPC specific. When RPC operations
occur, metric events about those operations will be generated and reported to
provide insight into those operations. By adding RPC labels to metric events
it allows for finely tuned filtering.
```

File: specification/metrics/semantic_conventions/rpc.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
+# General RPC conventions
+
```
Comment (arminru) [inline specification/metrics/semantic_conventions/rpc.md:2]: It would be nice if you could add a TOC here (have a look at some of the other documents, these make use of `markdown-toc` for that).

**Review by arminru — COMMENTED** [review]
Please add an entry in CHANGELOG.md for your new semantic convention. Thanks!

**Comment by arminru** [comment]
@jsuereth The build checks required to pass for merging are currently saying "Waiting for status to be reported" because we switched to Github actions in the meantime. If you merge or rebase onto the current master, those changes should be picked up and the build should pass.

File: specification/metrics/semantic_conventions/rpc.md (hunk 11 above)
Comment (jsuereth) [reply specification/metrics/semantic_conventions/rpc.md:2]: added

File: specification/metrics/semantic_conventions/rpc.md (hunk 9 above)
Comment (jsuereth) [reply specification/metrics/semantic_conventions/rpc.md:10]: removed

File: specification/metrics/semantic_conventions/rpc.md (hunk 7 above)
Comment (jsuereth) [reply specification/metrics/semantic_conventions/rpc.md:48]: It is

File: specification/metrics/semantic_conventions/rpc.md (hunk 8 above)
Comment (jsuereth) [reply specification/metrics/semantic_conventions/rpc.md:73]: see above

**Comment by jsuereth** [comment]
> Please add an entry in CHANGELOG.md for your new semantic convention. Thanks!

Done!
